### PR TITLE
test the review controller

### PR DESCRIPTION
**Pull request by ranilo**
Added test for the reviews controller.
Checking the index get returns a json.
Checking a new review post updates the restaurant rating.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -41,6 +41,7 @@ gem 'bootsnap', '>= 1.1.0', require: false
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
+  gem "factory_bot_rails"
```
Comment (ozbendavid) [inline Gemfile]: please use ' for strings with no interpulation

File: app/controllers/reviews_controller.rb [hunk 2]
```diff
@@ -25,18 +25,21 @@ def create
     @review = Review.new(review_params)
 
       if @review.save
```
Comment (ozbendavid) [inline app/controllers/reviews_controller.rb:27]: indentation

File: Gemfile [hunk 3]
```diff
@@ -58,6 +60,7 @@ group :test do
   gem 'selenium-webdriver'
   # Easy installation and use of chromedriver to run system tests with Chrome
   gem 'chromedriver-helper'
+  gem 'rspec-rails', '~> 3.8'
```
Comment (ooffeerr) [inline Gemfile:64]: Weird. Are you running Selenium tests (Ones that instrument a browser with clicks and gestures?)

File: app/controllers/reviews_controller.rb [hunk 4]
```diff
@@ -25,18 +25,21 @@ def create
...
+    else
+      render json:{}, status:403
       end
     end
 
   # PATCH/PUT /reviews/1
   # PATCH/PUT /reviews/1.json
   def update
       if @review.update(review_params)
+        update_rating restaurant: @restaurant
```
Comment (ozbendavid) [inline app/controllers/reviews_controller.rb]: Should we check if @restaurant exists?

File: app/controllers/reviews_controller.rb [hunk 5]
```diff
@@ -50,11 +53,23 @@ def destroy
     render status: 200, json: @review
   end
 
-  def update_rating
-    size = @restaurant.reviews.length
-    updated = size == 0 ? 0 : @restaurant.reviews.to_a.sum{|review| review.rating} / size
-    @restaurant.rating = updated
-    @restaurant.save
+  def update_rating (params)
```
Comment (ozbendavid) [inline app/controllers/reviews_controller.rb]: Controllers should stay slim
I would move it to the Restaurant model and call it by
@restaurant.update_rating

File: app/controllers/reviews_controller.rb [hunk 6]
```diff
@@ -25,18 +25,21 @@ def create
...
 
       if @review.save
-        update_rating
+        update_rating restaurant: @restaurant
         render json: @review
       else
         render json: @review.errors
       end
+    else
+      render json:{}, status:403
```
Comment (ooffeerr) [inline app/controllers/reviews_controller.rb]: `Rack::Utils.status_code(:forbidden)` would be slightly more readable.

File: app/controllers/reviews_controller.rb [hunk 7]
```diff
@@ -50,11 +53,23 @@ def destroy
...
 
-  def update_rating
-    size = @restaurant.reviews.length
-    updated = size == 0 ? 0 : @restaurant.reviews.to_a.sum{|review| review.rating} / size
-    @restaurant.rating = updated
-    @restaurant.save
+  def update_rating (params)
+      restaurant = Restaurant.find(params[:restaurant][:id])
+      size = restaurant.reviews.length
+      updated = size == 0 ? 0 : restaurant.reviews.to_a.sum{|review| review.rating} / size
```
Comment (ozbendavid) [inline app/controllers/reviews_controller.rb]: Keep descriptive naming, something like updated_rating
Also, instead of update rating and do save, you can use the update method
restaurant.update(rating: updated_rating)

File: spec/reviews_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+FactoryBot.define do
+  factory :restaurant do
+  end
+end
+
+
+describe  ReviewsController , :type => :controller do
+  context "GET index" do
```
Comment (ozbendavid) [inline spec/reviews_controller_spec.rb]: Use ' ' for simple strings

File: spec/reviews_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+FactoryBot.define do
+  factory :restaurant do
```
Comment (ozbendavid) [inline spec/reviews_controller_spec.rb]: All the factories of a specific class should be in a different file

File: spec/reviews_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+      expect(response.content_type).to eq "application/json"
+    end
+  end
+
+  context 'POST a new review' do
+    let(:restaurant) { FactoryBot.create(:restaurant)}
+    let(:rating) {3}
+
+    it "create a new review" do
+      comment_str = "test comment"
```
Comment (ozbendavid) [inline spec/reviews_controller_spec.rb]: Instead of separating the review and the rating I would define a review factory (in a different file) and use Faker to generate rating and comment.
Then, in the post-call bellow you can user review.to_json instead of writing field by field

File: spec/reviews_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  factory :restaurant do
+  end
+end
+
+
+describe  ReviewsController , :type => :controller do
+  context "GET index" do
+    it "has a 200 status code" do
+      get :index
+      expect(response.status).to eq(200)
```
Comment (ozbendavid) [inline spec/reviews_controller_spec.rb]: can be more descriptive as
expect(response).to have_http_status(200)

File: spec/reviews_controller_spec.rb (hunk 8 above)
Comment (ozbendavid) [inline spec/reviews_controller_spec.rb]: Also, I see that the convention is that action should be with # e.g. 'GET #index'

File: spec/reviews_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+FactoryBot.define do
+  factory :restaurant do
+  end
+end
+
+
+describe  ReviewsController , :type => :controller do
+  context "GET index" do
+    it "has a 200 status code" do
```
Comment (ozbendavid) [inline spec/reviews_controller_spec.rb]: I like to split the tests to the smallest pieces possible

for example :
it 'has a 200 status code' do ...
it 'returns json'
it 'returns an array of reviews' <-- This is a missing test - Create Reviews using factories and see that the index return this list

File: spec/reviews_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    end
+  end
+
+  context 'POST a new review' do
+    let(:restaurant) { FactoryBot.create(:restaurant)}
+    let(:rating) {3}
+
+    it "create a new review" do
+      comment_str = "test comment"
+      post :create, :params => {:review => {:name=>"test rev", :comment=>comment_str, :rating=>rating, :restaurant_id=>restaurant.id}}
```
Comment (ozbendavid) [inline spec/reviews_controller_spec.rb]: params: review.to_json

File: spec/reviews_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+      expect(response.status).to eq(200)
```
Comment (ozbendavid) [inline spec/reviews_controller_spec.rb]: Again, splitting to small tests
it 'has a 200 status code'
it 'returns the new review'

I would not test the new rating here...since we can count on the ActiveRecord mechanism to add this review to the restaurant we intended

Since it is better to move the update_rating to the Restuarant model, this is where I would have tested this function

File: spec/reviews_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+      expect(response.body).to include(comment_str)
+      expect(controller.update_rating restaurant: restaurant).to eq rating
+    end
```
Comment (ozbendavid) [inline spec/reviews_controller_spec.rb:41]: Also need to write negative test of post

**Review by ozbendavid — CHANGES_REQUESTED** [review]

File: app/controllers/reviews_controller.rb [hunk 16]
```diff
@@ -50,11 +53,23 @@ def destroy
...
+      restaurant.rating = updated
+      restaurant.save
+    restaurant.rating
```
Comment (ooffeerr) [inline app/controllers/reviews_controller.rb]: Is this indentend differently for a reason?

File: app/controllers/reviews_controller.rb (hunk 7 above)
Comment (ooffeerr) [reply app/controllers/reviews_controller.rb]: +1 to Oz's thoughts, plus - I useally like to hide code that requires effort to read behind methods, so it's reader might be spared of the gorry details - in this case, the whole average calculation could hide under a 'calculate average' method.

File: app/controllers/reviews_controller.rb [hunk 17]
```diff
@@ -50,11 +53,23 @@ def destroy
...
+
+  end
+
+
+  def render_json (params)
+    if(params[:content].nil?)
+      render status:200, json: {}
```
Comment (ooffeerr) [inline app/controllers/reviews_controller.rb]: Again, you could use the `status_code` as suggested above.

File: spec/reviews_controller_spec.rb (hunk 8 above)
Comment (ooffeerr) [reply spec/reviews_controller_spec.rb]: The reason being - "" strings are being interpulated, as they might contain a variable or some lazy evaluated expression, where '' strings are considered as-is values - hence you'll sparing so computation from the ruby interperter.

File: spec/reviews_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+FactoryBot.define do
+  factory :restaurant do
+  end
+end
+
+
+describe  ReviewsController , :type => :controller do
```
Comment (ooffeerr) [inline spec/reviews_controller_spec.rb:3]: What would you expected the GET of a ReviewsController to **behave** like?
Is it returning some specific value? Does it assume some kind of state for returning this value (Like, having some reviews submitted before - some you can create with the proper factory?)
Maybe it changes a preexisting state?
These would be the tests you should try to write - not sure if it's within the context of the pet project, but it might be a nice practice.

File: spec/reviews_controller_spec.rb [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+      expect(response.status).to eq(200)
+      expect(response.content_type).to eq "application/json"
+    end
+  end
+
+  context 'POST a new review' do
+    let(:restaurant) { FactoryBot.create(:restaurant)}
+    let(:rating) {3}
+
+    it "create a new review" do
```
Comment (ooffeerr) [inline spec/reviews_controller_spec.rb]: The test's title claims to verify a review was _created_, but claim isn't reflected at the code (Or am I missing it?)
I'd expected a new review to be persisted (i.e `expect(Review.last.rating).to eq(:rating)`)

**Review by ooffeerr — COMMENTED** [review]
Noice!
Added my some of my thoughts.

File: Gemfile (hunk 3 above)
Comment (ranilo) [reply Gemfile:64]: nope, probably added by the use of scaffold??

File: .rspec [hunk 20]
```diff
@@ -0,0 +1 @@
+--require spec_helper
```
Comment (odedmagger) [inline .rspec:1]: Interesting!

File: app/controllers/restaurants_controller.rb [hunk 21]
```diff
@@ -49,7 +49,7 @@ def update
   # DELETE /restaurants/1.json
   def destroy
     @restaurant.destroy
-    render status: 200, json: @restaurant
+    render status: Rack::Utils.status_code(:ok), json: @restaurant
```
Comment (odedmagger) [inline app/controllers/restaurants_controller.rb:52]: If we just destroyed `@restaurant`, why are we passing `@restaurant` to the rener call? Sounds dangerous and potentially harmful.
It's sufficient to keep an empty body here.
Also, if you're not doing anything beyond the default behavior,, I would check whether the base class' implementation of `destroy` does what you need. Would spare you the need to override it.

File: app/controllers/restaurants_controller.rb [hunk 22]
```diff
@@ -49,7 +49,7 @@ def update
   # DELETE /restaurants/1.json
   def destroy
     @restaurant.destroy
```
Comment (odedmagger) [inline app/controllers/restaurants_controller.rb:51]: Why did you need to save restaurant as a data member?

File: app/controllers/reviews_controller.rb [hunk 23]
```diff
@@ -1,12 +1,12 @@
 class ReviewsController < ApplicationController
   before_action :set_review, only: [:show, :edit, :update, :destroy]
-  before_action :set_restaurant, only: [:create]
+  before_action :set_restaurant, only: [:create, :update]
 
   # GET /reviews
   # GET /reviews.json
   def index
     @reviews = Review.all
```
Comment (odedmagger) [inline app/controllers/reviews_controller.rb:8]: Why do you need to store reviews on a data member?

File: app/controllers/reviews_controller.rb [hunk 24]
```diff
@@ -50,23 +54,31 @@ def destroy
...
 
-  def update_rating
-    size = @restaurant.reviews.length
-    updated = size == 0 ? 0 : @restaurant.reviews.to_a.sum{|review| review.rating} / size
-    @restaurant.rating = updated
-    @restaurant.save
+
+  def render_json (params)
+    if (params[:content].nil?)
+      render status: Rack::Utils.status_code(:ok), json: {}
```
Comment (odedmagger) [inline app/controllers/reviews_controller.rb:60]: I think `render status: :ok` works well here.

File: app/controllers/reviews_controller.rb [hunk 25]
```diff
@@ -21,26 +21,30 @@ def edit
...
 
       if @review.save
-        update_rating
         render json: @review
       else
         render json: @review.errors
       end
-      end
+    else
+      render json: {:error=>"restaurant #{review_params[:restaurant_id]} dose not exist"}, status: Rack::Utils.status_code(:forbidden)
```
Comment (odedmagger) [inline app/controllers/reviews_controller.rb:33]: The `:x => y` syntax is old Ruby syntax and we are trying to avoid it.

The new syntax should look more familiar to you :) - `{ error: 'whatever whatever' }`

File: app/models/restaurant.rb [hunk 26]
```diff
@@ -16,4 +16,12 @@
 class Restaurant < ApplicationRecord
   has_many :reviews
 
+  def update_rating
+    reload
```
Comment (odedmagger) [inline app/models/restaurant.rb:20]: Why is this needed?

File: app/models/review.rb [hunk 27]
```diff
@@ -13,5 +13,17 @@
 
 class Review < ApplicationRecord
   belongs_to :restaurant
+  before_save :clean_comments
+  after_save :update_rating
```
Comment (odedmagger) [inline app/models/review.rb:17]: Let me offer an alternative:
Right now, you're working hard to update the restaurant rating. You do it from the review (which by good design principles should not be aware of the restaurant object if it doesn't have to), and it's entirely possibly you're missing a flow where the rating should be updated but it isn't right now.

The alternative would be to calculate the restaurant rating 'on the fly' whenever it's needed, rather than have restaurant store the rating and update it on changes.

This approach has its pros and cons - it's your call whether to use it. Would love to hear what did you decide.

File: app/models/review.rb [hunk 28]
```diff
@@ -13,5 +13,17 @@
...
+  after_destroy :update_rating
+
+  def update_rating
+    restaurant.update_rating
+  end
+
+  def clean_comments
+    self.comment = comment.gsub!(/[^0-9A-Za-z]/, '')
+
```
Comment (odedmagger) [inline app/models/review.rb:26]: Remove newline

File: app/models/review.rb [hunk 29]
```diff
@@ -13,5 +13,17 @@
...
   belongs_to :restaurant
+  before_save :clean_comments
+  after_save :update_rating
+  after_destroy :update_rating
+
+  def update_rating
+    restaurant.update_rating
+  end
+
+  def clean_comments
```
Comment (odedmagger) [inline app/models/review.rb:24]: Why is this needed?

File: spec/factories/restaurant.rb [hunk 30]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :restaurant do
+    name{Faker::Name.name}
```
Comment (odedmagger) [inline spec/factories/restaurant.rb:3]: space after name

File: spec/rails_helper.rb [hunk 31]
```diff
@@ -0,0 +1,63 @@
+# This file is copied to spec/ when you run 'rails generate rspec:install'
```
Comment (odedmagger) [inline spec/rails_helper.rb:1]: This boilerplate code shouldn't have been part of the PR.

File: spec/spec_helper.rb [hunk 32]
```diff
@@ -0,0 +1,96 @@
+# This file was generated by the `rails generate rspec:install` command. Conventionally, all
```
Comment (odedmagger) [inline spec/spec_helper.rb:1]: Boilerplate code - shouldn't have been part of the PR.

File: app/controllers/reviews_controller.rb (hunk 25 above)
Comment (odedmagger) [reply app/controllers/reviews_controller.rb:33]: Also:
- dose --> does
- simply use :forbidden instead of that Rack:: thing.

File: app/controllers/reviews_controller.rb [hunk 33]
```diff
@@ -21,26 +21,30 @@ def edit
...
     end
+  end
 
   # PATCH/PUT /reviews/1
   # PATCH/PUT /reviews/1.json
   def update
+    if @restaurant
```
Comment (odedmagger) [inline app/controllers/reviews_controller.rb:40]: For checking existence, use `.present?`. e.g. `@restaurant.present?`
And yes, this can safely be called on null. In Ruby, null is an object too.

File: app/controllers/reviews_controller.rb [hunk 34]
```diff
@@ -50,23 +54,31 @@ def destroy
...
   end
 
   private
-    def set_review
-      @review = Review.find(params[:id])
-    end
 
-    def set_restaurant
+  def set_review
+    @review = Review.find(params[:id])
```
Comment (odedmagger) [inline app/controllers/reviews_controller.rb:69]: Why is this needed?
An alternative is to simply call `Review.find(x)` whenever you need to get a review, without storing it to a data member.

We should store stuff on data members when we need it to persist across method calls, and to 'actually store it on the object'.

File: app/models/review.rb [hunk 35]
```diff
@@ -13,5 +13,17 @@
...
+    self.comment = comment.gsub!(/[^0-9A-Za-z]/, '')
+
+  end
```
Comment (odedmagger) [inline app/models/review.rb:28]: Remove newline

File: spec/factories/restaurant.rb [hunk 36]
```diff
@@ -0,0 +1,10 @@
...
+    cuisine {Faker::Name.name}
+    rating {0}
+    address {Faker::Address}
+    max_delivery_time {Faker::Number.rand}
+    business_friendly {Faker::Boolean}
```
Comment (odedmagger) [inline spec/factories/restaurant.rb:8]: What's business-friendly?

File: spec/factories/restaurant.rb [hunk 37]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :restaurant do
+    name{Faker::Name.name}
+    cuisine {Faker::Name.name}
+    rating {0}
+    address {Faker::Address}
+    max_delivery_time {Faker::Number.rand}
```
Comment (odedmagger) [inline spec/factories/restaurant.rb:7]: What range does this random number give? What range do you want max_delivery_time to be? Is max_delivery_time discrete or continuous?

File: spec/factories/reviews.rb [hunk 38]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :random_review, class: Review do
```
Comment (odedmagger) [inline spec/factories/reviews.rb:2]: Sufficient to call it 'Review'. As long as whoever uses this factory does not explicitly define its fields, it is a convention for those fields to be populated by arbitrary or random legal values.

File: spec/reviews_controller_spec.rb [hunk 39]
```diff
@@ -0,0 +1,55 @@
+require 'rails_helper'
+
+describe  ReviewsController , :type => :controller do
```
Comment (odedmagger) [inline spec/reviews_controller_spec.rb:3]: Use `type: controller` instead of old syntax

File: spec/reviews_controller_spec.rb [hunk 40]
```diff
@@ -0,0 +1,55 @@
...
+  context 'POST a review for a missing restaurant' do
+    let(:review) { FactoryBot.build_stubbed :random_review, restaurant_id: 403 }
+
+    it 'has a 200 status code' do
+      post :create, params:  {review: review.as_json}
+      expect(response).to have_http_status(:forbidden)
+    end
+  end
+
+
```
Comment (odedmagger) [inline spec/reviews_controller_spec.rb:44]: Remove newline

File: spec/reviews_controller_spec.rb [hunk 41]
```diff
@@ -0,0 +1,55 @@
+require 'rails_helper'
+
+describe  ReviewsController , :type => :controller do
+  context 'GET #index' do
```
Comment (odedmagger) [inline spec/reviews_controller_spec.rb:4]: Use `describe` and note `context`.
Remove the `'GET `' prefix.
Read here why: http://www.betterspecs.org/

I won't add similar comments in other relevant lines within this spec, pretend they're there :-) .

File: spec/reviews_controller_spec.rb [hunk 42]
```diff
@@ -0,0 +1,55 @@
...
+    it 'has a 200 status code' do
+      get :index
+      expect(response).to have_http_status(:success)
+    end
```
Comment (odedmagger) [inline spec/reviews_controller_spec.rb:8]: add newline after end.

File: spec/reviews_controller_spec.rb (hunk 18 above)
Comment (odedmagger) [reply spec/reviews_controller_spec.rb:3]: 👍

File: spec/reviews_controller_spec.rb [hunk 43]
```diff
@@ -0,0 +1,55 @@
...
+    it 'returns json' do
+      get :index
+      expect(response.content_type).to eq 'application/json'
+    end
+  end
+
+  context 'POST a new review' do
+    let(:restaurant) { FactoryBot.create(:restaurant)}
+    let(:review) { FactoryBot.build_stubbed :random_review, restaurant_id: restaurant.id }
```
Comment (odedmagger) [inline spec/reviews_controller_spec.rb:17]: Why are we using `build_stubbed`? How does it differ from using build and create?

File: spec/reviews_controller_spec.rb [hunk 44]
```diff
@@ -0,0 +1,55 @@
...
+    let(:restaurant) {FactoryBot.create(:restaurant)}
+
+     it 'check restaurant review updated' do
+       review = FactoryBot.create(:random_review, restaurant_id: restaurant.id, rating: 2)
+       expect(restaurant.reload.rating).to eq review.rating
+     end
+  end
+
+
+end
```
Comment (odedmagger) [inline spec/reviews_controller_spec.rb:55]: add newline after end

File: spec/reviews_controller_spec.rb [hunk 45]
```diff
@@ -0,0 +1,55 @@
...
+
+  context 'adding a review' do
+    let(:restaurant) {FactoryBot.create(:restaurant)}
+
+     it 'check restaurant review updated' do
+       review = FactoryBot.create(:random_review, restaurant_id: restaurant.id, rating: 2)
+       expect(restaurant.reload.rating).to eq review.rating
+     end
+  end
+
```
Comment (odedmagger) [inline spec/reviews_controller_spec.rb:53]: Remove unecessary newlines

File: spec/reviews_controller_spec.rb [hunk 46]
```diff
@@ -0,0 +1,55 @@
...
+      post :create, params:  {review: review.as_json}
+      expect(response).to have_http_status(:forbidden)
+    end
+  end
+
+
+  context 'adding a review' do
+    let(:restaurant) {FactoryBot.create(:restaurant)}
+
+     it 'check restaurant review updated' do
```
Comment (odedmagger) [inline spec/reviews_controller_spec.rb:48]: This is one out of several scenarios related to rating we should probably test.

Please add more tests covering other possible scenarios.

File: spec/reviews_controller_spec.rb [hunk 47]
```diff
@@ -0,0 +1,55 @@
...
+       review = FactoryBot.create(:random_review, restaurant_id: restaurant.id, rating: 2)
```
Comment (odedmagger) [inline spec/reviews_controller_spec.rb:49]: You can pass `restaurant: restaurant` instead of passing restaurant id, assuming that in the model you defined the right associations (has_many, belongs_to).

**Review by odedmagger — CHANGES_REQUESTED** [review]
Glad to see your WeEat shaping up :) .

Don't worry about the sheer amount of comments - the project is a great opportunity to familiarize yourself with the little details of RoR development.

I also suggest reaching out to Paz Aricha from Yaron's squad to review this; Paz has volunteered to review WeEat server-side PRs.

As for the requested changed, don't feel obliged to do anything, but these are the main comments:
- Saving variables as controller's data members - let's do it only when there's a good reason.
- returning a variable right after calling destory over it - let's not do that, it's totally OK for a DELETE call to return empty body.
- Use `present?` when checking existence (similar methods you should know and love: .nil? , .blank? , .empty? and .any? )
- Add more tests to ensure the rating calculation behaves OK on all foreseeable scenarios.

I also would like to consider calculating rating on the fly rather than storing the rating.

File: app/controllers/reviews_controller.rb (hunk 24 above)
Comment (ranilo) [reply app/controllers/reviews_controller.rb:60]: this will not result in html.erd file rendering then?

File: app/models/restaurant.rb (hunk 26 above)
Comment (ranilo) [reply app/models/restaurant.rb:20]: The restaurant record is in memory when creating a new review, (for checking it exists)
therefor the update is needed so the restaurant will hold the reference to the new review..

File: app/models/review.rb (hunk 27 above)
Comment (ranilo) [reply app/models/review.rb:17]: Calculating the rating as the reviews accumulate will allow a short "get" for the user view.
I am not sure yet about the performance of the DB, yet.
But this is not a big issue in the case of we eat!...

File: app/models/review.rb (hunk 29 above)
Comment (ranilo) [reply app/models/review.rb:24]: Comments can contain null characters, this is for cleaning them out.

File: spec/rails_helper.rb (hunk 31 above)
Comment (ranilo) [reply spec/rails_helper.rb:1]: ok

File: app/controllers/reviews_controller.rb (hunk 34 above)
Comment (ranilo) [reply app/controllers/reviews_controller.rb:69]: Is this a better practice then using the before_action :set_review, function?

File: spec/factories/restaurant.rb (hunk 36 above)
Comment (ranilo) [reply spec/factories/restaurant.rb:8]: 10 bis

File: spec/factories/reviews.rb (hunk 38 above)
Comment (ranilo) [reply spec/factories/reviews.rb:2]: I did not use random numbers to limit the rating to be 0-3 stars, for preventing the default values from randomly failing. 
For negative tests, will specify numbers not in the rang.

File: spec/reviews_controller_spec.rb (hunk 42 above)
Comment (ranilo) [reply spec/reviews_controller_spec.rb:8]: added by atom on save

File: spec/reviews_controller_spec.rb (hunk 43 above)
Comment (ranilo) [reply spec/reviews_controller_spec.rb:17]: faster, no db access (like build)  
build_stabbed makes objects look like they’ve been persisted by creating associations.
